### PR TITLE
Add if present

### PR DESCRIPTION
**Pull request by stackmystack**

### CODE REVIEW

File: test/with_ar/all_agnostic_test.rb [hunk 1]
```diff
@@ -95,7 +95,9 @@ def teardown
       end
 
       def t(scope, node)
-        scope.select(node.as('res')).to_a.first.res
+        res = scope.select(node.as('res'))
+        # puts res.to_sql
+        res.to_a.first.res
       end
```
Comment (akimd) [inline test/with_ar/all_agnostic_test.rb:101]: Did you really mean to keep this part?

File: lib/arel_extensions/visitors/mysql.rb [hunk 2]
```diff
@@ -252,7 +252,18 @@ def visit_ArelExtensions_Nodes_RegexpReplace o, collector
...
+        # One use case we met is
+        # `case…when…then(valid_date).else(Arel.null).format(…)`.
+        #
+        # In this case, `o.col_type` is `nil` but we have a legitimate type in
+        # the expression to be formated.  The following is a best effort to
+        # infer the proper type.
+        type =
+          o.col_type.nil? && !o.expressions[0].return_type.nil? \
+          ? o.expressions[0].return_type \
+          : o.col_type
```
Comment (akimd) [inline lib/arel_extensions/visitors/mysql.rb:264]: I'm not sure I understand what's going on here.  To me "NULL in, NULL out", and it looks like you're trying to do something else.

File: test/with_ar/all_agnostic_test.rb (hunk 1 above)
Comment (stackmystack) [reply test/with_ar/all_agnostic_test.rb:101]: Damn git …

File: lib/arel_extensions/visitors/mysql.rb (hunk 2 above)
Comment (stackmystack) [reply lib/arel_extensions/visitors/mysql.rb:264]: Why does it look like I am trying to do something else?

I am doing a best effort to deduce the effective type of the expression. 

When we use `Arel.when(…).then(…).format(…)`: `col_type` _will be_ `nil`. And when that happens, then no formatting will be applied to the `expression`, unless we take some extra steps:

- If it's in `col_type`, then we're good to go.
- If `col_type` is `nil` and the `expression`'s return type isn't `nil`, then we use that.
- In all other cases, we want to fallback on `col_type`.

Comment (akimd) [reply lib/arel_extensions/visitors/mysql.rb:264]: Ok, now I understand what's going on, thanks.  Please make sure to include a test case for this situation (I can't see one).
And BTW, "formatted", not "formated" (in the comment above the lines we're commenting about).

Comment (akimd) [reply lib/arel_extensions/visitors/mysql.rb:264]: > for this situation
I meant a `when` that evaluates to `null`.

Comment (stackmystack) [reply lib/arel_extensions/visitors/mysql.rb:264]: The test case is included here https://github.com/Faveod/arel-extensions/pull/92/files#diff-94c8fa6ad26fb0fa6a96fc1dc185da1db8515b21993915cc530a436acce79d2fR775

Yeah it's implicit and buried. You prefer a separate test case?

Comment (akimd) [reply lib/arel_extensions/visitors/mysql.rb:264]: We are not referring to the same case.  I really meant "evaluates to NULL".  So I expect an `assert_nil`.

**Review by jdelporte — COMMENTED** [review]
the ``.else(Arel.null)`` is not necessary

**Pull request merged**